File: Java/src/gov/nasa/larcfm/ACCoRD/AlertingMofN.java
/*
 * Implementation of M of N algorithm for alerting
 * Contact: Cesar A. Munoz
 * Organization: NASA/Langley Research Center
 *
 * Copyright (c) 2011-2017 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */

package gov.nasa.larcfm.ACCoRD;

import java.util.LinkedList;
import java.util.Queue;

public class AlertingMofN {

	private int m_;
	private int n_;
	private int max_;
	private Queue<Integer> queue_;

	/* 
	 * Creates an empty M of N object
	 */
	public AlertingMofN() {
		m_ = 1;
		n_ = 0;
		max_ = 0;
		queue_ = new LinkedList<Integer>();
	}
	
	/*
	 * Creates a M of N object for given values of m, n, and maximum alert level. 
	 * Assumes  m >= 1 and m <= n
	 */
	public AlertingMofN(int m, int n, int max_alert_level) {
		m_ = m;
		n_ = n;
		max_ = max_alert_level;
		queue_ = new LinkedList<Integer>();
		for (int i=0;i<n_;++i) {
			queue_.add(0);
		}
	}

	/*
	 * Set new values for m and n.
	 * Assumes  m >= 1 and m <= n
	 */
	public void setMofN(int m, int n) {
		m_ = m;
		n_ = n;
		reset();
	}

	/*
	 * Set new values for maximum alert level
	 * Assumes max_alert_level >= 1
	 */
	public void setMaxAlertLevel(int max_alert_level) {
		max_ = max_alert_level;
		reset();
	}

	/*
	 * Reset M of N object.
	 */
	public void reset() {
		queue_.clear();
		for (int i=0;i<n_;++i) {
			queue_.add(0);
		}
	}

	/* 
	 * Return M of N value for a given alert level
	 */
	public int m_of_n(int alert_level) {
		if (queue_.isEmpty() || alert_level < 0 || alert_level > max_ || m_ <= 1 || max_ <= 0) {
			return alert_level;
		} 
		queue_.poll();
		queue_.add(alert_level);
		int[] count = new int[max_];
		for (int i=0; i < max_; ++i) {
			count[i]=0;
		}
		for (int alert : queue_) {
			for (int i=alert-1;i >= 0;--i) {
				count[i]++;
			}
		}	
		for (int i=max_-1; i >= 0; --i) {
			if (count[i] >= m_) {
				return i+1;
			}
		}
		return 0;
	}

}

File: Modules/ACCoRD/src/AlertingMofN.cpp
/*
 * Implementation of M of N algorithm for alerting
 * Contact: Cesar A. Munoz
 * Organization: NASA/Langley Research Center
 *
 * Copyright (c) 2011-2018 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */

#include "AlertingMofN.h"
#include "Util.h"

#include <vector>

namespace larcfm {

/*
 * Creates an empty M of N object
 */
AlertingMofN::AlertingMofN() {
  m_ = 0;
  n_ = 0;
  hysteresis_time_ = 0;
  persistence_time_ = 0;
  reset();
}

/*
 * Creates a M of N object for given values of m, n
 * Assumes  m >= 1 and m <= n
 */
AlertingMofN::AlertingMofN(int m, int n, double hysteresis_time, double persistence_time) {
  reset(m,n,hysteresis_time,persistence_time);
}

/*
 * Reset M of N object.
 */
void AlertingMofN::reset() {
  _init_time_ = NaN;
  _last_time_ = NaN;
  _alert_ = -1;
  _max_ = 0;
  _queue_.clear();
  for (int i=0;i<n_;++i) {
    _queue_.push_back(0);
  }
}

/*
 * Creates a M of N object for given values of m, n
 * Assumes  m >= 1 and m <= n
 */
void AlertingMofN::reset(int m, int n, double hysteresis_time, double persistence_time) {
  m_ = m;
  n_ = n;
  hysteresis_time_ = Util::max(hysteresis_time,persistence_time);
  persistence_time_ = persistence_time;
  reset();
}

/*
 * Set new values for m and n.
 * Assumes  m >= 1 and m <= n
 */
void AlertingMofN::setMofN(int m, int n) {
  if (m_ != m || n_ != n) {
    m_ = m;
    n_ = n;
    reset();
  }
}

bool AlertingMofN::isValid() const {
  return m_ > 0 && m_ <= n_;
}

/*
 * Return M of N value for a given alert level
 */
int AlertingMofN::m_of_n(int alert_level) {
  if (_queue_.empty() || alert_level < 0 ||  !isValid()) {
    return alert_level;
  }
  if (alert_level > _max_) {
    _max_ = alert_level;
  }
  _queue_.pop_front();
  _queue_.push_back(alert_level);
  if (_max_ == 0) {
    return 0;
  }
  std::vector<int> count(_max_,0);
  for (std::deque<int>::const_iterator alert_ptr = _queue_.begin(); alert_ptr != _queue_.end(); ++alert_ptr) {
    for (int i=(*alert_ptr)-1;i >= 0;--i) {
      count[i]++;
    }
  }
  for (int i=_max_-1; i >= 0; --i) {
    if (count[i] >= m_) {
      return i+1;
    }
  }
  return 0;
}

/*
 * Return M of N value for a given alert level at a current time.
 * In addition of m_of_n, this also applies hysteresis and persistence
 */
int AlertingMofN::m_of_n(int alert_level, double current_time) {
  if (!ISNAN(_last_time_)) {
    if (current_time == _last_time_) {
      return _alert_;
    }
    if (current_time < _last_time_ || current_time - _last_time_ > hysteresis_time_) {
      reset();
    }
  }
  int alert_mofn = m_of_n(alert_level);
  if (alert_mofn < 0) {
    _last_time_ = current_time;
    _alert_ = alert_mofn;
    return alert_mofn;
  }
  if (!ISNAN(_init_time_) && _alert_ > 0 && alert_mofn < _alert_ && current_time >= _init_time_ &&
      current_time - _init_time_ < persistence_time_) {
    // Do nothing. Keep the previous alert_
  } else {
    if (alert_mofn > 0 && alert_mofn != _alert_) {
      _init_time_ = current_time;
    }
    _alert_ = alert_mofn;
  }
  _last_time_ = current_time;
  return _alert_;
}

} /* namespace larcfm */
